Caching

Tyrrrz opened the pull request:
Closes #557

bruno-garcia commented on src/Sentry/SentryOptionsExtensions.cs:
```diff
@@ -182,6 +183,15 @@ public static SentryOptions UseStackTraceFactory(this SentryOptions options, ISe
+        /// <summary>
+        /// Sets the cache directory.
+        /// </summary>
+        public static SentryOptions SetCacheDirectory(this SentryOptions options, string directoryPath)
```
What about we hold only a `string` in the options (to make it easier for folks to add the path, remove this method) and we create the DirectoryInfo internally?

Tyrrrz replied:
I kinda like the explicitness of `DirectoryInfo` specifically for user-facing interface. I wouldn't use it internally.

bruno-garcia commented on src/Sentry/Internal/CachingBackgroundWorker.cs:
```diff
@@ -0,0 +1,89 @@
+                    cancellationToken
+                ).ConfigureAwait(false);
+
+                await Transport.SendEnvelopeAsync(envelope, cancellationToken).ConfigureAwait(false);
```
We default to deleting the file. If there was an exception while trying to send it, for example. Or if Sentry returned 400, 500.

The exception is when we can clearly identify we couldn't connect to Sentry (timeout, or offline).
In this case we keep the file around to try again later.

That said I believe we're missing a mechanism to communicate that here. And some try/finally(delete)

Tyrrrz replied:
That's true, I will update it

bruno-garcia commented on src/Sentry/Internal/CachingBackgroundWorker.cs:
```diff
@@ -0,0 +1,89 @@
+            Envelope envelope,
+            CancellationToken cancellationToken = default)
+        {
+            await FlushCacheAsync(cancellationToken).ConfigureAwait(false);
```
Does it mean we're not storing the `envelope` received here until we deal with cached files?
I believe we should cache the first first of all, to guarantee in the event of a failure (device our of power, process killed, crashed) we didn't lose the event.

The behavior for mobile at least, is to alwauys write to disk first, then try to send. and delete if it succeeded. Also delete it if we got a failure that could happen again once we retry.

Tyrrrz replied:
Good point, will update.

bruno-garcia commented on src/Sentry/Internal/CachingBackgroundWorker.cs:
```diff
@@ -0,0 +1,89 @@
+
+        private async ValueTask CacheEnvelopeAsync(Envelope envelope, CancellationToken cancellationToken = default)
+        {
+            if (GetCachedEnvelopeFiles().Count() >= 50)
```
30 I believe is the general default, this will need to propagate from Options.

```suggestion
            if (GetCachedEnvelopeFiles().Count() >= 30)
```

Tyrrrz replied:
Yeah, I just put a random number for now

bruno-garcia commented on src/Sentry/Internal/CachingBackgroundWorker.cs:
```diff
@@ -0,0 +1,89 @@
+            var envelopeFile = new FileInfo(
+                Path.Combine(
+                    _cacheDirectory.FullName,
+                    $"{DateTimeOffset.Now}_{envelope.TryGetEventId()}_{envelope.GetHashCode()}")
```
You used a `.envelope` extension above, we need it here?

What about UtcNow so the file names are not subjective to daylight savings/timezone changes?

What if not event id in the envelope, maybe `?? ...`

Tyrrrz replied:
Good points. I didn't think this through very much yet.

Tyrrrz replied:
Fixed.

If event ID is not set then it will be empty, so it will look like this:

```
123456789__91231293.envelope
```

Note the double underscore. Since the file names are not really meant to be read by users, I think it's okay. It will still be sorted by date and combination of unix time + hash code will ensure there are no collisions. Event ID is there for debugging purposes, mostly.

bruno-garcia commented on src/Sentry/SentryOptions.cs:
```diff
@@ -360,6 +361,12 @@ public bool Debug
+        /// Path to the directory where events are stored locally for resilience purposes.
+        /// If set to <code>null</code>, caching will not be used.
+        /// </summary>
+        public DirectoryInfo? CacheDirectory { get; set; }
```
I think `string` is probably a better type, to for example bind from configuration files

Tyrrrz replied:
Hm, I didn't think about configuration files. That may indeed be a good reason to change it to `string`. Ah, unfortunate that C# discourages proper static typing so much.

bruno-garcia reviewed (commented):
I see it's a draft, but a first pass hopefully some points will be useful

Tyrrrz commented:
@bruno-garcia addressed the issues you pointed out.  Going to start working on tests soon, so if there are any immediate design concerns, we should address them now :)

bruno-garcia commented on src/Sentry/Internal/BackgroundWorkerBase.cs:
```diff
@@ -144,7 +125,7 @@ private async ValueTask WorkerAsync()
                         catch (Exception exception)
                         {
-                            _options.DiagnosticLogger?.LogError(
+                            Options.DiagnosticLogger?.LogError(
```
Seems the format on this one expects 3 arguments but if the first is Exception (not part of `params`) are we missing one?

Tyrrrz replied:
Can't figure out what it's missing :/
Probably some info got lost in transition from events to envelopes. I'll remove the extra placeholder for now.

bruno-garcia commented on src/Sentry/Internal/CachingBackgroundWorker.cs:
```diff
@@ -0,0 +1,146 @@
+        }
+
+        private IEnumerable<FileInfo> GetEnvelopeFiles() =>
+            _cacheDirectory.Exists
```
Should we check if it exists in the `ctor` and do some LogWarn or LogError?
I'm afraid this will be hard to debug

Tyrrrz replied:
That would probably be a bad idea. File system is very volatile, so even if it existed during construction, it may get deleted somehow after that, but before writing files. We should always treat working with FS as if it's in a constantly unpredictable state.

bruno-garcia replied:
I agree with everything you said. But I would expect to see some logs if the directory isn't there when the user explicitly set a path and expects offline caching to be working,

Tyrrrz replied:
Do you think it's unusual for someone to set a caching directory to something that doesn't exist? I personally would expect it to be normal.

I can add a log when we save a file, but the directory is not there.

bruno-garcia commented on src/Sentry/Internal/CachingBackgroundWorker.cs:
```diff
@@ -0,0 +1,146 @@
+                        envelope.TryGetEventId()
+                    );
+                }
+                catch (IOException ex)
```
I'm afraid we could have some I/O error that's not really recoverable and we'll be retrying that forever. Maybe we need to be more strict on what we consider OK to retry

Tyrrrz replied:
Unfortunately most transient errors we would want to retry on just throw base `IOException`. Some have inner exceptions (like `SocketException`) but not all and it's probably implementation-dependent. I guess we'd have to parse exception messages if we really wanted to be selective.

bruno-garcia replied:
That's an option. For now we're good with IOException, the client returning 500 wouldn't throw so at least we know that if it hits Sentry it'll be deleted

bruno-garcia commented on src/Sentry/Internal/CachingBackgroundWorker.cs:
```diff
@@ -0,0 +1,146 @@
+                    // i.e. loss of connection, failure to connect, etc.
+                    // Instead break to not violate the order and wait until
+                    // the next batch to attempt sending again.
+                    Options.DiagnosticLogger?.Log(
+                        SentryLevel.Error,
```
```suggestion
                    Options.DiagnosticLogger?.LogError(
```

bruno-garcia commented on src/Sentry/Internal/CachingBackgroundWorker.cs:
```diff
@@ -0,0 +1,146 @@
+                    // indicate a successfully completed, but error response.
+                    envelopeFile.Delete();
+
+                    Options.DiagnosticLogger?.Log(
+                        SentryLevel.Error,
```
```suggestion
                    Options.DiagnosticLogger?.LogError(
```

bruno-garcia commented on src/Sentry/Internal/CachingBackgroundWorker.cs:
```diff
@@ -0,0 +1,146 @@
+            CancellationToken cancellationToken = default)
+        {
+            // Add this envelope to cache
+            await StoreEnvelopeAsync(envelope, cancellationToken).ConfigureAwait(false);
```
If we throw here, the worker thread above is supposed to log and continue, right?
I wonder if we'll drop the item and deal with it "gracefully".
We need to make sure we don't end up retrying something broken in a tight loop here on a worker thread or killing that thread and stopping all event submission

Tyrrrz replied:
Yeah, this method is called from the worker which has like 5 layers of `try`/`catch` :)
How do you think the workflow should look like?

bruno-garcia replied:
> Yeah, this method is called from the worker which has like 5 layers of try/catch :)


Does this mean the worker won't crash? I believe that's the case.

Tyrrrz replied:
Yeah it won't crash.

bruno-garcia commented on src/Sentry/Internal/CachingBackgroundWorker.cs:
```diff
@@ -0,0 +1,146 @@
+
+namespace Sentry.Internal
+{
+    internal class CachingBackgroundWorker : BackgroundWorkerBase
```
We're not going to be able to use this class later outside of the background thread..
What if we used composition here somehow?

The reason I ask is the Xamarin crash handler use case, where we need to write to disk but not sure we can do on a threadpool thread.

We can try that first before changing anything, and we can change it later too, just thinking ahead a bit

Tyrrrz replied:
We may want to consolidate into one approach somehow, because otherwise we'll have to have two caching background workers, one for Xamarin and one for everything else.

bruno-garcia replied:
I'll try your code now on Xamarin crash handler

bruno-garcia reviewed (commented):
Looking awesome!

Tyrrrz commented on src/Sentry/Internal/BackgroundWorkerBase.cs:
```diff
@@ -8,65 +8,42 @@
 namespace Sentry.Internal
 {
-    internal class BackgroundWorker : IBackgroundWorker, IDisposable
+    internal abstract class BackgroundWorkerBase : IBackgroundWorker, IDisposable
```
I find that the `Base` suffix is much more common for abstract classes that have base functionality, while abstract classes without functionality (basically interfaces) don't have that suffix. For example, `ControllerBase` from aspnetcore mvc.

Tyrrrz commented:
Updated @bruno-garcia.

Dispose on `BackgroundWorker` is so messy, I didn't bother trying to dispose the `ITransport` in it for the fear of introducing bugs. I had some major refactor done for `BackgroundWorker` in one of my previous attempts, but had to discard it because I switched to a different approach now.

Tyrrrz commented:
Work of art now.

Tyrrrz commented:
Hm, the tests are failed. This must be the work of an enemy stand. I'll see what's wrong.

lucas-zimerman commented:
@Tyrrrz I think the windows issue is related to the one that's happening with the UWP PR

Tyrrrz commented:
Added blocking on SentrySdk.Init

bruno-garcia commented on test/Sentry.Tests/Internals/Http/CachingTransportTests.cs:
```diff
@@ -0,0 +1,227 @@
+
+            // (transport stops failing)
+            innerTransport.ClearReceivedCalls();
+            isFailing = false;
```
Can u make sure it didn't do the job yet?

```suggestion
            _ = innerTransport.Received(0).SendEnvelopeAsync(Arg.Any<Envelope>(), Arg.Any<CancellationToken>());
            isFailing = false;
```

Tyrrrz replied:
well it did call that method when trying (and failing) initially

bruno-garcia commented on test/Sentry.Tests/Internals/Http/CachingTransportTests.cs:
```diff
@@ -0,0 +1,227 @@
+
+        public void Dispose()
+        {
+            if (Directory.Exists(CacheDirectoryPath))
+            {
+                Directory.Delete(CacheDirectoryPath, true);
+            }
+        }
```
It just noops if there's no folder to delete

```suggestion
            Directory.Delete(CacheDirectoryPath, true);
        }
```

Tyrrrz replied:
No it doesn't 😄

bruno-garcia commented on src/Sentry/Internal/Lock.cs:
```diff
@@ -0,0 +1,21 @@
+
+        public Lock() => _signal = new Signal(true);
+
+        public async ValueTask<IDisposable> ClaimAsync(CancellationToken cancellationToken = default)
```
```suggestion
        public async ValueTask<IDisposable> AcquireAsync(CancellationToken cancellationToken = default)
```

bruno-garcia commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,266 @@
+            {
+                foreach (var filePath in Directory.EnumerateFiles(_processingDirectoryPath))
+                {
+                    Directory.CreateDirectory(_cacheDirectoryPath);
```
This could be created outside of the `foreach`.
It should no-op if it already exists.

bruno-garcia commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,266 @@
+            // Processing directory may already contain some files left from previous session
+            // if the worker has been terminated unexpectedly.
+            // Move everything from that directory back to cache directory.
+            if (Directory.Exists(_processingDirectoryPath))
```
`_cacheDirectoryPath` is a subdirectory of `_cacheDirectoryPath` so creating `_cacheDirectoryPath` in the foreach isn't needed.

bruno-garcia commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,266 @@
+            return targetFilePath;
+        }
+
+        private async ValueTask ProcessCacheAsync(CancellationToken cancellationToken = default)
```
Please make this the first method (or at least `private` method) since this is what we should be reading first.

bruno-garcia commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,266 @@
+            using var lockClaim = await _cacheDirectoryLock.ClaimAsync(cancellationToken).ConfigureAwait(false);
+
+            // If over capacity - remove oldest envelope file
+            while (
+                GetCacheFilePaths().Count() >= _options.MaxQueueItems &&
+                GetCacheFilePaths().FirstOrDefault() is { } oldestEnvelopeFilePath)
```
What about assigning the result of `GetCacheFilePaths` before the `FirstOrDefault`?
Maybe a local function would help

bruno-garcia commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,266 @@
+
+            // Tell the worker that there is work available
+            // (file stream MUST BE DISPOSED prior to this)
+            _workerSignal.Release();
```
If these operations above fail, we don't `Release`?
Are we missing a try finally here?

Tyrrrz replied:
No. Why would we need to release it in that case?

bruno-garcia commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,266 @@
+        {
+            // Stop worker and wait until it finishes
+            _workerCts.Cancel();
+            await _worker.ConfigureAwait(false);
```
Since we call this from the `Dispose` method, awaiting on a Task that fails throws. Should we try/catch here?

lucas-zimerman commented on src/Sentry/SentryOptions.cs:
```diff
@@ -51,6 +52,7 @@ public class SentryOptions : IScopeOptions
         /// </summary>
         internal ISdkIntegration[]? Integrations { get; set; }
 
+
```
Is this extra line required?

bruno-garcia commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,266 @@
+        }
+
+        // This method asynchronously blocks until the envelope is written to cache, but not until it's sent
+        public async ValueTask SendEnvelopeAsync(
```
This should come right after the constructor

Tyrrrz replied:
Why?

lucas-zimerman commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,266 @@
+
+                using (var envelopeFile = File.OpenRead(envelopeFilePath))
+                using (var envelope =
+                    await Envelope.DeserializeAsync(envelopeFile, cancellationToken).ConfigureAwait(false))
```
If the file got corrupted, isn't deserialization going to throw an Exception and the remaining Envelopes are not going to be loaded?

Tyrrrz replied:
Good point.

Tyrrrz commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,271 @@
+            // Processing directory may already contain some files left from previous session
+            // if the worker has been terminated unexpectedly.
+            // Move everything from that directory back to cache directory.
+            if (Directory.Exists(_processingDirectoryPath))
```
Are you sure it does? I'm pretty sure I remember it throwing in the past.

Tyrrrz replied:
![image](https://user-images.githubusercontent.com/1935960/99104235-68c0b700-25e9-11eb-88ed-3b6f134780fe.png)

Tyrrrz replied:
https://docs.microsoft.com/en-us/dotnet/api/system.io.directory.enumeratefiles?view=net-5.0

lucas-zimerman commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,266 @@
+
+            using (var stream = File.Create(envelopeFilePath))
+            {
+                await envelope.SerializeAsync(stream, cancellationToken).ConfigureAwait(false);
```
If an IOException like Disk full happens, it'll not release the workerSignal?

Tyrrrz replied:
No, it won't.

Tyrrrz commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,266 @@
+        {
+            StopWorkerAsync().GetAwaiter().GetResult();
+
+            _workerSignal.Dispose();
```
Dispose inner transport

Tyrrrz commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,266 @@
+            _innerTransport = innerTransport;
+            _options = options;
+
+            _cacheDirectoryPath = !string.IsNullOrWhiteSpace(options.CacheDirectoryPath)
```
Use a subdirectory: `{cacheDir}/Sentry/DSN`

Tyrrrz commented:
Mention: multiple instances of the same apps are not supported

bruno-garcia commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,266 @@
+                    }
+                    catch (Exception ex)
+                    {
+                        _options.DiagnosticLogger?.LogError(
```
Since this is potentially dangerous (we will throw on the next iteration) lets slow things down:

```csharp
await Task.Delay(500);
```

Tyrrrz commented on src/Sentry/Internal/Http/CachingTransport.cs:
```diff
@@ -0,0 +1,271 @@
+
+            var targetFilePath = Path.Combine(_processingDirectoryPath, Path.GetFileName(filePath));
+
+            Directory.CreateDirectory(_processingDirectoryPath);
```
Yes, it's not a guarantee, but it's rather cheap if the directory already exists and it's not like we're running in a tight loop anyway. This gives a bit more resiliency because we don't care about previous state.

Tyrrrz commented on test/Sentry.Tests/Internals/Http/CachingTransportTests.cs:
```diff
@@ -0,0 +1,227 @@
+    public class CachingTransportTests : IDisposable
+    {
+        private string CacheDirectoryPath { get; } = Path.Combine(
+            Directory.GetCurrentDirectory(),
```
This lets us inspect the actual cache directory more easily. Plus I don't like polluting system directories if I have a choice.

Tyrrrz replied:
Note: with the current setup, the cache directory ends up inside `bin/debug` directory in tests.

lucas-zimerman replied:
https://github.com/lucas-zimerman/ContribSentryDotNet/blob/master/ContribSentry.Testing/TempFolder.cs

Something like this would be nice for testing the cached folder, it'll delete the files once the test is finished with or without an exception.

Tyrrrz replied:
That does not work if the test is canceled by xUnit timeout, but I think it looks nicer so I'll refactor to that

Tyrrrz commented:
@bruno-garcia should we keep the retry logic, given your experimentation with Xamarin?

bruno-garcia commented on src/Sentry/Internal/Signal.cs:
```diff
@@ -0,0 +1,39 @@
+                }
+            }
+
+            _semaphore.Release();
```
racey, sync the whole thing to make sure no 2 threads reach Release
